lib/onelogin/ruby-saml/logoutresponse.rb

```diff
@@ -69,46 +104,211 @@ def status_code
+      def validate(soft = true, request_id = nil, get_params = nil)
+        @errors = []
+        valid_state?(soft)                      &&
+        validate_success_status(soft)           &&
+        validate_structure(soft)                &&
+        valid_in_response_to?(soft, request_id) &&
+        validate_destination(soft)              &&
+        valid_issuer?(soft)                     &&
+        validate_signature(soft, get_params)
+      end
+
       private
 
+      # Validates that the Logout Response provided in the initialization is not empty, 
+      # also check that the setting and the IdP cert were also provided 
+      # If fails, the error is added to the errors array.
+      # @param soft [Boolean] soft Enable or Disable the soft mode (In order to raise exceptions when the logout response is invalid or not)
+      # @return [Boolean] True if the required info is found, otherwise False if soft=True
+      # @raise [ValidationError] if soft == false and validation fails
+      #
       def valid_state?(soft = true)
-        if response.empty?
-          return soft ? false : validation_error("Blank response")
+        if response.nil? or response.empty?
+          error_msg = "Blank Logout Response"
+          @errors << error_msg
+          return soft ? false : validation_error(error_msg)
         end
 
         if settings.nil?
-          return soft ? false : validation_error("No settings on response")
+          error_msg = "No settings on Logout Response"
+          @errors << error_msg
+          return soft ? false : validation_error(error_msg)
         end
 
         if settings.issuer.nil?
-          return soft ? false : validation_error("No issuer in settings")
+          error_msg = "No issuer in settings"
+          @errors << error_msg
+          return soft ? false : validation_error(error_msg)
         end
 
         if settings.idp_cert_fingerprint.nil? && settings.idp_cert.nil?
-          return soft ? false : validation_error("No fingerprint or certificate on settings")
+          error_msg = "No fingerprint or certificate on settings"
+          @errors << error_msg
+          return soft ? false : validation_error(error_msg)
+        end
+
+        true
+      end
+
+      # Validates the Logout Response against the specified schema.
+      # If fails, the error is added to the errors array
+      # @param soft [Boolean] soft Enable or Disable the soft mode (In order to raise exceptions when the logout response is invalid or not)
+      # @return [Boolean] True if the XML is valid, otherwise False if soft=True
+      # @raise [ValidationError] if soft == false and validation fails
+      #
+      def validate_structure(soft = true)
+        begin 
+          valid = valid_saml?(document, soft)
+          unless valid
+            @errors << "Invalid Logout Response. Not match the saml-schema-protocol-2.0.xsd"
+          end
+          valid
+        rescue OneLogin::RubySaml::ValidationError => e
```

Comment:
While I realize that this pattern may have been here before this PR was opened, IME using exceptions in a validation context leads to bad JuJu. An error state is not necessarily exceptional, it is only exceptional if you have *not* accounted for the case.

This may be a comment outside the context of this PR.